cep-recipe-jinja.md

```diff
@@ -0,0 +1,511 @@
+
+### The `pin_compatible` function
+
+Pin compatible will pin the dependency to the same version as "previously"
+resolved in the `host` or `build` environment. This is useful to ensure that the
+same package is used at run time as was used at build time.
+
+Example:
+
+```yaml
+requirements:
+  host:
+    - numpy
+  run:
+    - ${{ pin_compatible('numpy', exact=True) }}
+    # or alternatives
+    # - ${{ pin_compatible('numpy', lower_bound='x.x.x', upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound=None, upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound="1.0", upper_bound='x') }}
+    # - ${{ pin_compatible('numpy', lower_bound="1.0", upper_bound="2.0") }}
+```
+
+### The `pin_subpackage` function
+
+Pin subpackage will pin the dependency to the same version as another
+sub-package from the recipe (or the current package itself). This is useful to
+ensure that multiple outputs from a recipe are linked together or to export the
+correct `run_exports` for a package.
+
+Example:
+
+```yaml
+outputs:
+  - package:
+      name: libfoo
+      version: "1.2.3"
+  - package:
+      name: foo
+      version: "1.2.3"
+    requirements:
+      run:
+        - ${{ pin_subpackage('libfoo', exact=True) }}
+```
+
+## The `match` function
+
+The `match` function is used to match a variant with a version spec. It returns
+`true` if the version spec matches the variant and `false` otherwise.
+
+For example, it can be used in the following way:
+
+```yaml
+requirements:
+  - $ {{ "six" if match(python, "<3.8") }}
+  - $ {{ "six" if match(python, "3.8") }}
+  - $ {{ "six" if match(python, "==3.8") }}
+  - $ {{ "six" if match(python, "3.8.*") }}
+  - $ {{ "six" if match(python, ">=3.8,<3.10") }}
+```
+
+In this case the value from the `python` _variant_ is used to add or remove
+optional dependencies. Note that generalizes and replaces selectors from old
+recipes, such as `# [py38]` or `# [py3k]`.
+
+The version comparison rules follow those of the `conda` version comparison
+rules.
+
+## The `hash` variable
+
+`${{ hash }}` is the variant hash and is useful in the build string computation.
+This used to be `PKG_HASH` in the old recipe format. Since the `hash` variable
+depends on the variant computation, it is only available in the `build.string`
+field and is computed after the entire variant computation is finished.
+
+## The `env` object
+
+The `env` object is used to retrieve environment variables and inject them into
+the recipe. There are two ways to do this:
+
+- `${{ env.get("MY_ENV_VAR") }}` will return the value of the environment
+  variable `MY_ENV_VAR` or throw an error if the environment variable is not
+  set.
+- `${{ env.get("MY_ENV_VAR", default="default_value") }}` will return the value
+  of the environment variable `MY_ENV_VAR` or `"default_value"` if it is unset.
+
+You can also check for the existence of an environment variable:
+
+- `${{ env.exists("MY_ENV_VAR") }}` will return a boolean `true` if the
```

Comment:
Is the `.exists` a standard Jinja method?
(Sorry, I didn't look this up. Just want to make sure we are conscious this in case it's some custom thing vs a plain Python  `"MY_ENV_VAR" in env`.)

Reply:
No, it's not standard jinja. You're right, we might be able to implemnetn the functions to make `"FOO" in env` work, but I am not sure if its worth it (also seems a bit more magical?).

We also don't use a regular "dict" or "sequence" object to be able to throw proper errors etc.

Reply:
Ah, I also missed that `env.get("MY_ENV_VAR")` should throw an error in case `MY_ENV_VAR` isn't set -- meaning I totally missed that `env` isn't meant to be a `dict`.
TBH, I don't have a strong opinion then (I actually like a method better; no idea what people would prefer syntax- and/or naming-wise).

Reply:
`get` just returns `None` for undefined keys in Python and Jinja

https://docs.python.org/3/library/stdtypes.html#dict.get